core/src/main/java/org/elasticsearch/common/io/stream/NamedWriteableRegistry.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.io.stream;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Registry for {@link NamedWriteable} objects. Allows to register and retrieve prototype instances of writeable objects
+ * given their name.
+ */
+public class NamedWriteableRegistry {
+
+    private final Map<String, NamedWriteable<?>> registry = new HashMap<>();
+
+    public static final NamedWriteableRegistry EMPTY = new NamedWriteableRegistry();
+
+    /**
+     * Registers a {@link NamedWriteable} prototype given its category
+     */
+    public synchronized void registerPrototype(String category, NamedWriteable<?> namedWriteable) {
+        String key = getKey(category, namedWriteable.getWriteableName());
+        if (registry.containsKey(key)) {
+            throw new IllegalArgumentException("named writeable of type [" + namedWriteable.getClass().getName() + "] with name [" + namedWriteable.getWriteableName() + "] " +
+                    "is already registered by type [" + registry.get(key).getClass().getName() + "] under the [" + category + "] category");
+        }
+        registry.put(key, namedWriteable);
+    }
+
+    /**
+     * Returns a prototype of the {@link NamedWriteable} object identified by the name provided as argument and its category
+     */
+    public synchronized <C> NamedWriteable<C> getPrototype(String category, String name) {
+        String key = getKey(category, name);
+        @SuppressWarnings("unchecked")
+        NamedWriteable<C> namedWriteable = (NamedWriteable<C>) registry.get(key);
+        if (namedWriteable == null) {
+            throw new IllegalArgumentException("unknown named writeable with name [" + name + "] and category [" + category + "]");
+        }
+        return namedWriteable;
```

Comment:
We need to cast indeed, but I want to give the compiler opportunities to find errors, which is never possible when one starts definiting methods whose generic parameter is only used in the return value. By the way I'm thinking that we could make casts more safe by making category a class instead of a string, and this class would be the base class of the object that the namedwriteables can deserialize